unimol_tools/unimol_tools/models/unimolv2.py

```diff
@@ -0,0 +1,587 @@
+        x = mul * x.unsqueeze(-1) + bias
+        x = x.expand(-1, -1, -1, self.K)
+        mean = self.means.weight.float().view(-1)
+        std = self.stds.weight.float().view(-1).abs() + 1e-5
+        return gaussian(x.float(), mean, std).type_as(self.means.weight)
+    
+class NumericalEmbed(nn.Module):
+    """
+    Numerical embedding module, typically used for embedding edge features in graph neural networks.
+
+    Attributes:
+        - K: Output dimension for embeddings.
+        - mul, bias, w_edge: Embeddings for transformation parameters.
+        - proj: Projection layer to transform inputs.
+        - ln: Layer normalization.
+    """
+    def __init__(self, K=128, edge_types=1024, activation_fn='gelu'):
+        """
+        Initializes the NonLinearHead.
+
+        :param input_dim: The input dimension of the first layer.
+        :param out_dim: The output dimension of the second layer.
+        :param activation_fn: The activation function to use.
+        :param hidden: The dimension of the hidden layer; defaults to input_dim if not specified.
+        """
+        super().__init__()
+        self.K = K 
+        self.mul = nn.Embedding(edge_types, 1)
+        self.bias = nn.Embedding(edge_types, 1)
+        self.w_edge = nn.Embedding(edge_types, K)
+
+        self.proj = NonLinearHead(1, K, activation_fn, hidden=2*K)
+        self.ln = nn.LayerNorm(K)
+
+        nn.init.constant_(self.bias.weight, 0)
+        nn.init.constant_(self.mul.weight, 1)
+        nn.init.kaiming_normal_(self.w_edge.weight)
+
+
+    def forward(self, x, edge_type):    # edge_type, atoms
+        """
+        Forward pass of the NonLinearHead.
+
+        :param x: Input tensor to the classification head.
+
+        :return: The output tensor after passing through the layers.
+        """
+        mul = self.mul(edge_type).type_as(x)
+        bias = self.bias(edge_type).type_as(x)
+        w_edge = self.w_edge(edge_type).type_as(x)
+        edge_emb = w_edge * torch.sigmoid(mul * x.unsqueeze(-1) + bias)
+        
+        edge_proj = x.unsqueeze(-1).type_as(self.mul.weight)
+        edge_proj = self.proj(edge_proj)
+        edge_proj = self.ln(edge_proj)
+
+        h = edge_proj + edge_emb
+        h = h.type_as(self.mul.weight)
+        return h
+
+def molecule_architecture(model_size='84m'):
+    args = argparse.ArgumentParser()
+    if model_size == '84m':  
+        args.num_encoder_layers = getattr(args, "num_encoder_layers", 12)
+        args.encoder_embed_dim = getattr(args, "encoder_embed_dim", 768)
+        args.num_attention_heads = getattr(args, "num_attention_heads", 48)
+        args.ffn_embedding_dim = getattr(args, "ffn_embedding_dim", 768)
+    elif model_size == '164m':
+        args.num_encoder_layers = getattr(args, "num_encoder_layers", 24)
+        args.encoder_embed_dim = getattr(args, "encoder_embed_dim", 768)
+        args.num_attention_heads = getattr(args, "num_attention_heads", 48)
```

Comment:
`args.num_attention_heads` should be changed to `args.encoder_attention_heads` ?

I fail to load the 310m model weights because the MovementPredictionHead relies on `args.encoder_attention_heads`, and that is always 48 but it must be 64 for the 310m model.

Thank you for the awesome work!

Reply:
Thanks for your kindly reminder! Both ```args.num_attention_heads``` and ```args.encoder_attention_heads``` should be changed.